casper/src/main/resources/PoS.rhox

```diff
@@ -531,7 +547,86 @@ new  PoS,
+                TreeHashMap!("get", pendingRewardsMap, validatorKey, *currentRewardsMaybeCh) |
+                for(@currentRewardsMaybe <- currentRewardsMaybeCh) {
+                  match currentRewardsMaybe {
+                    Nil => currentRewardsCh!({})
+                    val => currentRewardsCh!(val)
+                  } |
+                  for (@currentRewards <- currentRewardsCh) {
+                    new mergeFunc, mergeResultCh in {
+                      // Merge current pending rewards with new pending rewards
+                      MergeMaps!(currentRewards, newRewards, *mergeFunc, *mergeResultCh) |
+                      // Sum values for current and new rewards
+                      contract mergeFunc(@v1, @v2, valMergeRet) = {
+                        valMergeRet!(v1 + v2)
+                      } |
+                      for (@mergeResult <- mergeResultCh) {
+                        // Update pending rewards state
+                        TreeHashMap!("set", pendingRewardsMap, validatorKey, mergeResult, return)
+                      }
+                    }
+                  }
+                }
+              }
+            } |
+
+            // Collect rewards from all validators / reset state to zero
+            contract PendingRewards(@"collect", @validatorKeys, @return) = {
+              new stCh, collectMaps, mapsListCh, mergedRewardsCh in {
+                // Get pending rewards (maps) for list of validators (public keys)
+                @ListOps!("fold", validatorKeys, [], *collectMaps, *mapsListCh) |
+                contract collectMaps(@validatorKey, @acc, mergeRet) = {
+                  new rewMapCh in {
+                    TreeHashMap!("get", pendingRewardsMap, validatorKey, *rewMapCh) |
+                    for (@rewMapMaybe <- rewMapCh) {
+                      new stateResetCh in {
+                        // Reset per validator pending rewards state
+                        TreeHashMap!("set", pendingRewardsMap, validatorKey, {}, *stateResetCh) |
+                        for (_ <- stateResetCh) {
+                          match rewMapMaybe {
+                            Nil => mergeRet!(acc)
+                            map => mergeRet!(acc ++ [map])
+                          }
+                        }
+                      }
+                    }
+                  }
+                } |
+
+                // Merge pending rewards (maps) to one map of pending rewards
+                for (@mapsList <- mapsListCh) {
+                  new collector in {
+                    @ListOps!("fold", mapsList, {}, *collector, return) |
+                    contract collector(@map, @acc, @mergeRet) = {
+                      new mergeFunc in {
+                        MergeMaps!(map, acc, *mergeFunc, mergeRet) |
+                        contract mergeFunc(@v1, @v2, valMergeRet) = {
+                          valMergeRet!(v1 + v2)
+                        }
```

Comment:
Merging of Int maps is the same as in "append", maybe to define contract for that?

Reply:
Replaced with `MergeIntMaps`, there is no need for more generalized version.